src/librustc_data_structures/obligation_forest/mod.rs

```diff
@@ -400,39 +566,72 @@ impl<O: ForestObligation> ObligationForest<O> {
     {
+        if self.offset.is_none() {
+            self.offset = Some(processor.register_variable_watcher());
+        }
         let mut errors = vec![];
         let mut stalled = true;
 
-        // Note that the loop body can append new nodes, and those new nodes
-        // will then be processed by subsequent iterations of the loop.
-        //
-        // We can't use an iterator for the loop because `self.nodes` is
-        // appended to and the borrow checker would complain. We also can't use
-        // `for index in 0..self.nodes.len() { ... }` because the range would
-        // be computed with the initial length, and we would miss the appended
-        // nodes. Therefore we use a `while` loop.
-        let mut index = 0;
-        while index < self.nodes.len() {
+        self.unblock_nodes(processor);
+
+        let nodes = &self.nodes;
+        self.unblocked.extend(
+            self.stalled_on_unknown
+                .drain(..)
+                .map(|index| Unblocked { index, order: nodes[index].node_number }),
+        );
+        while let Some(Unblocked { index, .. }) = self.unblocked.pop() {
+            // Skip any duplicates since we only need to processes the node once
+            if self.unblocked.peek().map(|u| u.index) == Some(index) {
+                continue;
+            }
+
             let node = &mut self.nodes[index];
 
+            if node.state.get() != NodeState::Pending {
+                continue;
+            }
+
             // `processor.process_obligation` can modify the predicate within
             // `node.obligation`, and that predicate is the key used for
             // `self.active_cache`. This means that `self.active_cache` can get
             // out of sync with `nodes`. It's not very common, but it does
             // happen, and code in `compress` has to allow for it.
-            if node.state.get() != NodeState::Pending {
-                index += 1;
-                continue;
+            let before = node.obligation.as_cache_key();
+            let result = processor.process_obligation(&mut node.obligation);
+            let after = node.obligation.as_cache_key();
+            if before != after {
+                node.alternative_predicates.push(before);
             }
 
-            match processor.process_obligation(&mut node.obligation) {
+            self.unblock_nodes(processor);
+            let node = &mut self.nodes[index];
+            match result {
                 ProcessResult::Unchanged => {
+                    let stalled_on = node.obligation.stalled_on();
+                    if stalled_on.is_empty() {
+                        // We stalled but the variables that caused it are unknown so we run
+                        // `index` again at the next opportunity
+                        self.stalled_on_unknown.push(index);
+                    } else {
+                        // Register every variable that we stalled on
```

Comment:
Just to make sure I'm following, a single obligation can definitely wind up on a list here multiple times, right? But only I suppose a fixed number in the end. i.e., if I have an obligation `Foo(?X, ?Y)`, I can be on the watch lists for both X and Y, and then when X is unified, I will still be blocked on Y, right?